---
title: "Clustering of Adult Zebrafish RGCs"
output: html_document
---

## Introduction
This notebook is intended to guide users through some of the major portions of analysis completed for clustering of adult zebrafish retinal ganglion cells. Most of the analysis was run using functionalites within the R pacakge Seurat (Satija et al.,Nature Biotechnology, 2015), an actively maintained set of tools for scRNA-seq analysis (https://satijalab.org/seurat/). `utilFxns.R` and `plottingFxns.R` are scripts containing custom scripts used for analysis and generating figures. These scripts are avaiable in the utils folder at  https://github.com/shekharlab/ZebrafishRGC. 

## Preprocessing

The first step is to load necessary packages and libraries.

```{r, warning=FALSE, message = FALSE}
library(Seurat)
library(dplyr)
library(reshape2)
library(ggplot2)
library(cowplot)
source("utils/utilFxns.R")
source("utils/plottingFxns.R")
```

Next, we create the initial Seurat object by loading the Count matrix. Count matrices are currently available at https://drive.google.com/drive/folders/1baRKtDkD4d8-6tG8P9v8VcjtUDpWeq5m?usp=sharing, but will later be avaiable from the Gene Expression Omnibus. The Count matrix is subsetted to remove a faulty sample and then used to create a Seurat object.

```{r, warning=FALSE, message=FALSE}
# Load adult zebrafish count matrix
Count.mat <- readRDS("../CountMatrices/ConsolidatedCounts_Zfish_013020.rds")

# Remove sample ZfishRGC16 due to wetting sample failure
cells.remove = grep("ZfishRGC16_",colnames(Count.mat), value=TRUE)
Count.mat = Count.mat[, setdiff(colnames(Count.mat), cells.remove)]

# Create Seurat object, removing genes that are expressed in fewer than 25 cells and removing cells
# that have fewer than 450 features
adult <- CreateSeuratObject(counts = Count.mat, project = "zfishRGC", min.cells = 25, min.features = 450)
```

Check quality metrics of the data, including RNA counts and mitochondrial scores.

```{r, warning=FALSE, message=FALSE}
adult[["percent.mt"]] <- PercentageFeatureSet(adult, pattern = "^MT-")
adult[["percent.rps"]] <- PercentageFeatureSet(adult, pattern = "^RPS")
adult[["percent.rpl"]] <- PercentageFeatureSet(adult, pattern = "^RPL")
adult[["percent.rp"]] <- adult[["percent.rps"]] + adult[["percent.rpl"]]

# Create Violin Plots of RNA counts, mitochondrial percentages, and ribosomal percentages
VlnPlot(adult, features = "nCount_RNA", pt.size = .3)
VlnPlot(adult, features = "percent.mt", pt.size = .3)
VlnPlot(adult, features = "percent.rp", pt.size = .3)
```

Set batch information.

```{r, warning=FALSE, message=FALSE}
# Change the order of factor
adult@meta.data$orig.ident = factor(adult@meta.data$orig.ident, levels = paste0("ZfishRGC",c(1:15)))

# Set batch information in meta.data
batchname = as.character(adult@meta.data$orig.ident)
batchid = rep("Batch0", length(batchname))
batchid[grep("ZfishRGC1$", batchname)] = "Batch1"
batchid[grep("ZfishRGC2", batchname)] = "Batch1"
batchid[grep("ZfishRGC3", batchname)] = "Batch2"
batchid[grep("ZfishRGC4", batchname)] = "Batch2"
batchid[grep("ZfishRGC5", batchname)] = "Batch3"
batchid[grep("ZfishRGC6", batchname)] = "Batch3"
batchid[grep("ZfishRGC7", batchname)] = "Batch3"
batchid[grep("ZfishRGC8", batchname)] = "Batch3"
batchid[grep("ZfishRGC9", batchname)] = "Batch4"
batchid[grep("ZfishRGC10", batchname)] = "Batch4"
batchid[grep("ZfishRGC11", batchname)] = "Batch5"
batchid[grep("ZfishRGC12", batchname)] = "Batch5"
batchid[grep("ZfishRGC13", batchname)] = "Batch5"
batchid[grep("ZfishRGC14", batchname)] = "Batch5"
batchid[grep("ZfishRGC15", batchname)] = "Batch5"
adult@meta.data$batch = factor(batchid)

table(adult@meta.data$orig.ident, adult@meta.data$batch)
```

## Clustering 

Using functionalities within Seurat, cluster the dataset. The data is first log-normalized and then all genes are scaled. The top 2000 variable genes are then detected and used to run PCA. The top 30 principle components were selected using an elbow plot to calculate cluster assignments.

```{r, warning=FALSE, message=FALSE}
# Log normalize the data, identify the top 2000 variable features, and scale all genes
adult <- NormalizeData(adult, normalization.method = "LogNormalize", scale.factor = 10000)
all.genes <- rownames(adult)
adult <- ScaleData(adult, features = all.genes)

# Run PCA on variable features and visualize the dimensionality of the dataset using an elbow plot
adult <- FindVariableFeatures(adult, selection.method = "vst", nfeatures = 2000)
adult <- RunPCA(adult, features = VariableFeatures(object = adult))

ElbowPlot(adult, ndims = 50)

# Compute clusters using 30 PCs and calculate UMAP and tSNE sembeddings
adult <- FindNeighbors(adult, dims = 1:30)
adult <- FindClusters(adult, resolution = .5)

adult <- RunTSNE(adult, dims = 1:30)
adult <- RunUMAP(adult, dims = 1:30)

# Visualize clusters
DimPlot(adult, reduction = "umap")
DimPlot(adult, reduction = "tsne")
```

Check for batch effects within the data by grouping the UMAP plot by batch.

```{r, warning=FALSE, message=FALSE,}
DimPlot(adult, reduction = "umap", group.by = "batch")
```

Perform data integration to correct for batch effects. 

```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Split object by batch number
adult.list <- SplitObject(adult, split.by = "batch")

# Normalize each dataset and find variable features
for (i in 1:length(adult.list)) {
  adult.list[[i]] <- NormalizeData(adult.list[[i]], verbose = FALSE)
  adult.list[[i]] <- FindVariableFeatures(adult.list[[i]], selection.method = "vst", 
                                             nfeatures = 1500, verbose = FALSE)
}

# Find Integration anchors
adult.anchors <- FindIntegrationAnchors(object.list = adult.list, dims = 1:40)

# Integrate Data
adult.integrated <- IntegrateData(anchorset = adult.anchors, dims = 1:40)
```

```{r, message = FALSE, error = FALSE, echo=FALSE}
# To save time and memory space , the above code is not run and the integrated object is loaded here instead.
adult.integrated <- readRDS("../Objects2020/zFish_SeuratClusteredIntegrated_02042020.rds")
```

Run dimensionality reduction and clustering on the new integrated data and visualize the effects of data integration.

```{r, message=FALSE, warning= FALSE, fig.width=12}
# Switch to integrated assay. The variable features of this assay are automatically set during IntegrateData
DefaultAssay(adult.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering
adult.integrated <- ScaleData(adult.integrated, verbose = FALSE)
adult.integrated <- RunPCA(adult.integrated, npcs = 40, verbose = FALSE)
adult.integrated <- RunUMAP(adult.integrated, reduction = "pca", dims = 1:40)
adult.integrated <- RunTSNE(adult.integrated, reduction = "pca", dims = 1:40)
adult.integrated <- FindNeighbors(adult.integrated, dims = 1:40)
adult.integrated <- FindClusters(adult.integrated) #Run Louvain algorithm

# Check out UMAP for the integrated vs. non-integrated data
p1 <- DimPlot(adult.integrated, reduction = "umap", group.by = "batch")
p2 <- DimPlot(adult, reduction = "umap", group.by = "batch")
plot_grid(p1, p2, labels = c("Integrated", "Non-integrated"))
```

Investigate clusters for canonical RGC markers and check RNA counts. 

```{r, fig.width= 10}
VlnPlot(adult.integrated,"nFeature_RNA", pt.size=0)
VlnPlot(adult.integrated,"rna_RBPMS2B", pt.size=0)
```


Check clusters that do not express RBPSM2B or have abnormal counts and remove contaminant cell types
```{r, message=FALSE, error=FALSE, eval=FALSE}
# 1 - Low # genes and no distinguishing markers
markers1 = FindMarkers(adult.integrated, ident.1 = 1, test.use="MAST", max.cells.per.ident = 1000)
# 3 - Muller Glia (RLBP1A, GLULA, APOEB)
markers3 = FindMarkers(adult.integrated, ident.1 = 3, test.use="MAST", max.cells.per.ident = 1000)
# 9 - Low RBPMS2B and rod markers
markers9 = FindMarkers(adult.integrated, ident.1 = 9, test.use="MAST", max.cells.per.ident = 1000)
# 13 - Glia, endothelial cells (PTGDS, APOEB, ZFP36L1B)
markers13=FindMarkers(adult.integrated, ident.1 = 13, test.use="MAST", max.cells.per.ident = 1000)
# 18 - Bipolar cells - VSX1, CAPB5A
markers18=FindMarkers(adult.integrated, ident.1 = 18, test.use="MAST", max.cells.per.ident = 1000)
# 20 - Amacrine cells - GAD1, GAD2, TFAP2A, SLC6A1B, GAD1B
markers20=FindMarkers(adult.integrated, ident.1 = 20, test.use="MAST", max.cells.per.ident = 1000)
# 22 - Low # genes, nothing distinguishing
markers22=FindMarkers(adult.integrated, ident.1 = 22, test.use="MAST", max.cells.per.ident = 1000)
# 30 - Abnormally high # genes, and distinguishing markers have a low effect size
markers30=FindMarkers(adult.integrated, ident.1 = 30, test.use="MAST", max.cells.per.ident = 1000)
# 39 - Glia doublet (RLBP1A, SEPP1A among top markers)
markers39=FindMarkers(adult.integrated, ident.1 = 39, test.use="MAST", max.cells.per.ident = 1000)
# 41 - Microglia (FCER1GL, CD74A)
markers41=FindMarkers(adult.integrated, ident.1 = 41, test.use="MAST", max.cells.per.ident = 1000)
# 42 - Endothelial cells (CLDN19, CD82A, CLDNK)
markers42=FindMarkers(adult.integrated, ident.1 = 42, test.use="MAST", max.cells.per.ident = 1000)
```

Remove contaminant cells from the dataset and subset nonintegrated object

```{r, message=FALSE, error=FALSE}
cells.remove = WhichCells(adult.integrated, idents= c(1,3,9, 13, 18, 20, 22, 30, 39, 41, 42))
adult = subset(adult,cells = setdiff(colnames(adult), cells.remove))
```

Since a large number of cells were removed, repeat integration and clustering steps.
```{r, message = FALSE, error= FALSE, eval = FALSE}
# Split by batch
adult.list <- SplitObject(adult, split.by = "batch")
# Normalize datasets and find variable features
for (i in 1:length(adult.list)) {
  print(i)
  adult.list[[i]] <- NormalizeData(adult.list[[i]], verbose = FALSE)
  adult.list[[i]] <- FindVariableFeatures(adult.list[[i]], selection.method = "vst", 
                                          nfeatures = 2000, verbose = FALSE)
}

# Find Integration anchors and integrate
adult.anchors <- FindIntegrationAnchors(object.list = adult.list, dims = 1:40)
adult.integrated <- IntegrateData(anchorset = adult.anchors, dims = 1:40)
DefaultAssay(adult.integrated) <- "integrated"

# Run the standard workflow for visualization and clustering using the integrated data
adult.integrated <- ScaleData(adult.integrated, verbose = FALSE)
adult.integrated <- RunPCA(adult.integrated, npcs = 40, verbose = FALSE)
adult.integrated <- RunUMAP(adult.integrated, reduction = "pca", dims = 1:40)
adult.integrated <- RunTSNE(adult.integrated, reduction = "pca", dims = 1:40)
adult.integrated <- FindNeighbors(adult.integrated, dims = 1:40)
adult.integrated <- FindClusters(adult.integrated) #Run Louvain algorithm
```

```{r, echo=FALSE}
# To save time and memory space , the above code is not run and the new integrated object is loaded here instead.
adult.integrated <- readRDS("../Objects2020/zFish_SeuratClusteredIntegratedPruned_Final_021820.rds")
Idents(adult.integrated) <- "integrated_snn_res.0.8"
```

Investigate DE features and remove any remanining contaminants.

```{r, message=FALSE, error=FALSE, eval=FALSE}
# Identify all DE genes across clusters
adult.markers <- FindAllMarkers(adult.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

# Cluster 10, 20, 22 lack unique differentially expressed genes
markers10=head(subset(adult.markers, cluster == 10))
markers20=head(subset(adult.markers, cluster == 20))
markers22=head(subset(adult.markers, cluster == 22))

# Cluster 35: Muller Glia (RLBP1A, APOEB)
markers35=head(subset(adult.markers, cluster == 35))
```

```{r, message=FALSE, error=FALSE}
# Remove contaminant cell types
Idents(adult.integrated) <- "integrated_snn_res.0.8"
idents.remove <- c(10, 20, 22, 35)

adult.integrated = subset(adult.integrated, idents = setdiff(levels(Idents(adult.integrated)), idents.remove))

# Refactor final cluster assignments as clusterID
adult.integrated@meta.data$clusterID = droplevels(Idents(adult.integrated))
levels(adult.integrated@meta.data$clusterID) = 1:length(levels(adult.integrated@meta.data$clusterID))
Idents(adult.integrated) <- "clusterID"
```

Construct dendrogram based on hierarchical clustering. Create an additional metadata column with the clusters ordered based on the dendrogram.

```{r, error=FALSE, message=FALSE, fig.height=10, fig.width=10}
# Build dendrogram
adult.integrated <- FindVariableFeatures(adult.integrated, selection.method = "vst", nfeatures = 500)
adult.integrated <- BuildClusterTree(adult.integrated)
PlotClusterTree(adult.integrated)
plot(adult.integrated@tools$BuildClusterTree)

# Reorder clusters according to dendrogram for dotplot plotting
tree_obj = adult.integrated@tools$BuildClusterTree
left_clusts = Seurat:::GetLeftDescendants(tree_obj, length(levels(adult.integrated@meta.data$clusterID))+1)
right_clusts = Seurat:::GetRightDescendants(tree_obj, length(levels(adult.integrated@meta.data$clusterID))+1)
clust_order = c(left_clusts, right_clusts)
adult.integrated@meta.data$dendro_order = factor(adult.integrated@meta.data$clusterID, levels = clust_order)
```

## Visualize the data

Visualize final cluster assignments on UMAP and tSNE

```{r, warning=FALSE, message=FALSE}
# Read in object, set order based on dendrogram, and convert genes to lower case
adult <- adult.integrated
Idents(adult) <- "dendro_order"
adult <- LowerCase_genes(adult, integration = TRUE)

# Visualize clusters on tSNE and UMAP
DimPlot(adult, reduction = "tsne", label = TRUE, group.by = "clusterID")
DimPlot(adult, reduction = "umap", label = TRUE, group.by = "clusterID")
```

Plot canonical RGC markers

```{r, error=FALSE, message=FALSE, fig.width=4, fig.height=6}
# Note, since some markers are not included in the top 2000 variable features of the integrated assay, the RNA assay is used instead
DotPlot(adult, features = c("robo2", "isl2b", "rbpms2b"), scale.min = 0, assay = "RNA") + RotatedAxis()
```

Identify all differentially expressed features

```{r, warning=FALSE, message=FALSE, eval=FALSE}
adult.markers <- FindAllMarkers(adult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```


```{r, warning=FALSE, message=FALSE, eval=FALSE, echo=FALSE}
adult.markers <- readRDS("../Objects2020/adult_ClusterMarkers_v2")
```

Plot differentially expressed genes across clusters

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=8}
mapGenes_vector <- vector()
for(i in levels(Idents(adult))){
  mapGenes_vector = union(mapGenes_vector, head(subset(adult.markers, cluster==i))[1:3,]$gene)
}
mapGenes_vector <- tolower(mapGenes_vector[2:length(mapGenes_vector)])
DotPlot(adult, features = mapGenes_vector, assay = "RNA") + RotatedAxis()
``` 

Plot relative size of each cluster

```{r, warning=FALSE, message=FALSE, fig.width=10}
PlotClusterSize(adult)
```